src/tooltips.ts

```diff
@@ -0,0 +1,244 @@
+	switch (placement) {
+		case 'right':
+			return rightArrowStyle;
+		case 'bottom':
+			return bottomArrowStyle;
+		case 'left':
+			return leftArrowStyle;
+		default:
+			return topArrowStyle;
+	}
+};
+
+export class Tooltips {
+	private tooltipElement: HTMLDivElement;
+	private tooltipElementInner: HTMLDivElement;
+	private tooltipElementArrow: HTMLDivElement;
+	private hideTimeout: number;
+	private initialised: boolean = false;
+
+	initialiseElements() {
+		if (this.initialised || !document || !document.body) {
+			return;
+		}
+
+		const tooltipElement = document.createElement('div');
+		tooltipElement.setAttribute(
+			'style',
+			`
+			display: block;
+			font-size: 12px;
+			font-style: normal;
+			font-weight: 400;
+			letter-spacing: normal;
+			line-break: auto;
+			line-height: 1.42857143;
+			padding: 5px;
+			position: absolute;
+			text-align: left;
+			text-align: start;
+			text-decoration: none;
+			text-shadow: none;
+			text-transform: none;
+			visibility: hidden;
+			white-space: normal;
+			word-break: normal;
+			word-spacing: normal;
+			word-wrap: normal;
+			z-index: 1070;
+		`,
+		);
+
+		const tooltipElementInner = document.createElement('div');
+		tooltipElementInner.setAttribute(
+			'style',
+			`
+			background: black;
+			border-radius: 4px;
+			color: white;
+			padding: 3px 8px;
+			textAlign: center;
+		`,
+		);
+
+		const tooltipArrow = document.createElement('div');
+		tooltipArrow.setAttribute(
+			'style',
+			`
+			${arrowStyle}
+			bottom: 0;
+			left: 50%;
+			margin-left: -5px;
+			border-width: 5px 5px 0;
+			border-top-color: #000;
+		`,
+		);
+
+		tooltipElement.appendChild(tooltipElementInner);
+		tooltipElement.appendChild(tooltipArrow);
+
+		document.body.appendChild(tooltipElement);
+
+		this.tooltipElement = tooltipElement;
+		this.tooltipElementInner = tooltipElementInner;
+		this.tooltipElementArrow = tooltipArrow;
+
+		this.initialised = true;
+	}
+
+	bindProps(props: any) {
+		if (props.tooltip) {
+			let trigger = 'hover';
+			let tooltipText: string;
+			let placement: Placement;
+			if (isString(props.tooltip)) {
+				tooltipText = props.tooltip;
+			} else {
+				tooltipText = props.tooltip.text;
+				placement = props.tooltip.placement;
+				if (props.tooltip.trigger) {
+					trigger = props.tooltip.trigger;
+				}
+			}
+			if (tooltipText) {
+				const showFn = (e: Event) => this.show(e, tooltipText, placement);
+				if (trigger === 'click') {
+					const oldFn = props.onClick || noop;
+					const hideFn = () => {
+						clearTimeout(this.hideTimeout);
+						this.hideTimeout = window.setTimeout(
+							() => this.hide(),
+							CLICK_TIMEOUT,
+						);
+					};
+					props.onClick = (e: Event) => showFn(e) || hideFn() || oldFn(e);
+				} else {
+					const oldMEFn = props.onMouseEnter || noop;
+					props.onMouseEnter = (e: Event) => showFn(e) || oldMEFn(e);
+
+					const oldMLFn = props.onMouseLeave || noop;
+					props.onMouseLeave = (e: Event) => this.hide() || oldMLFn(e);
```

Comment:
Here (and twice more above) there's bit of clever `() => a || b` to run `a` & then `b`. That works, but it's a little confusing and a little risky - if we accidentally return something truthy in any `a`, then `b` will never run.

Could these just be `() => { a(); b(); }` instead?